### PR TITLE
Export stock the beginnings

### PR DESCRIPTION
**Pull request by benrudolph**
@NoahCarnahan this PR is getting quite big in scope, so gonna make a breaking point here. i'm thinking it'd be really good to have some integration tests that create schemas from an app then actually export that data from inserted data

cc: @esoergel @biyeun 

:european_castle: or :fish:

### CODE REVIEW

File: corehq/apps/export/tests/test_table_configuration.py [hunk 1]
```diff
@@ -1,10 +1,24 @@
+from collections import namedtuple
+from mock import patch
 from django.test import SimpleTestCase
 
+from corehq.apps.commtrack.helpers import make_product
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_table_configuration.py]: F401 'make_product' imported but unused

File: corehq/apps/export/tests/test_table_configuration.py [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+from corehq.apps.products.models import Product
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_table_configuration.py]: F401 'Product' imported but unused

File: corehq/apps/export/models/new.py [hunk 3]
```diff
@@ -15,8 +17,12 @@
 )
 from corehq.apps.app_manager.models import Application
 from corehq.apps.app_manager.util import get_case_properties
+from corehq.apps.domain.models import Domain
+from corehq.apps.commtrack.models import StockState
```
Comment (dimagimon) [inline corehq/apps/export/models/new.py]: F401 'StockState' imported but unused

File: corehq/apps/export/tests/test_table_configuration.py [hunk 4]
```diff
@@ -343,3 +357,30 @@ def test_get_headers_with_extras(self):
...
+class StockExportColumnTest(SimpleTestCase):
+    domain = 'stock-domain'
+
+    def test_get_headers(self):
+        column = StockExportColumn(
+            domain=self.domain,
+            label="Stock",
+            item=ExportItem(),
+        )
+        get_ledger_path = 'corehq.form_processor.interfaces.dbaccessors.LedgerAccessors.get_ledger_values_for_product_ids'
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_table_configuration.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/export/models/new.py [hunk 5]
```diff
@@ -168,13 +168,7 @@ def create_default_from_export_item(table_path, item, app_ids_and_versions):
...
         }
 
-        if item.tag == PROPERTY_TAG_ROW:
-            column_class = RowNumberColumn
-            constructor_args["repeat"] = len([node for node in table_path if node.is_repeat])
-        else:
-            column_class = ExportColumn
-
-        column = column_class(**constructor_args)
+        column = ExportColumn(**constructor_args)
```
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: i think this should be `SplitExportColumn` when the item is multiple choice and expand questions is set to true

File: corehq/apps/export/models/new.py [hunk 6]
```diff
@@ -366,6 +366,9 @@ class ExportInstance(BlobMixin, Document):
     last_updated = DateTimeProperty()
     last_accessed = DateTimeProperty()
 
+    # Whether to add stock columns to the export
+    export_stock = BooleanProperty(default=False)
+
```
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: removed later

File: corehq/apps/export/models/new.py [hunk 7]
```diff
@@ -476,6 +503,7 @@ def __insert_system_properties(cls, table, properties, top=True):
 
         for static_column in properties:
             index, existing_column = table.get_column(static_column.item.path, static_column.item.transform)
+            (existing_column or static_column).extra_data(**data)
```
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: don't love this, but it's the only way to get more data into these columns like domain and repeat

File: corehq/apps/export/static/export/js/models.js [hunk 8]
```diff
@@ -181,9 +181,12 @@ Exports.ViewModels.ExportColumn.prototype.isCaseName = function() {
     return this.item.isCaseName();
 };
 
+Exports.ViewModels.ExportColumn.prototype.translatedHelp = function() {
```
Comment (dimagimon) [inline corehq/apps/export/static/export/js/models.js]: 'Exports' is not defined. (no-undef)

File: corehq/apps/export/static/export/js/models.js [hunk 9]
```diff
@@ -181,9 +181,12 @@ Exports.ViewModels.ExportColumn.prototype.isCaseName = function() {
...
+    return gettext(this.help_text);
```
Comment (dimagimon) [inline corehq/apps/export/static/export/js/models.js]: 'gettext' is not defined. (no-undef)

File: corehq/apps/export/models/new.py [hunk 10]
```diff
@@ -129,8 +135,9 @@ def get_value(self, doc, base_path, transform_dates=False, row_index=None):
         :param base_path:
         :return:
         """
-        # Confirm the ExportItem's path starts with the base_path
-        assert base_path == self.item.path[:len(base_path)]
+        if base_path != self.item.path[:len(base_path)]:
+            import ipdb; ipdb.set_trace()
```
Comment (dimagimon) [inline corehq/apps/export/models/new.py]: E702 multiple statements on one line (semicolon)

File: corehq/apps/export/static/export/js/models.js (hunk 8 above)
Comment (esoergel) [inline corehq/apps/export/static/export/js/models.js]: You can declare Exports as a global for this file by putting this at the top of the file:
```javascript
/*global Exports*/
```
as described [here](https://github.com/dimagi/js-guide/blob/master/linting.md#global-variables)

(also fine to just ignore for now)

File: corehq/apps/export/models/new.py [hunk 11]
```diff
@@ -450,24 +456,54 @@ def _insert_system_properties(cls, export_type, table):
...
             elif table.path == CASE_HISTORY_TABLE:
-                cls.__insert_system_properties(table, CASE_HISTORY_PROPERTIES)
+                cls.__insert_system_properties(table, CASE_HISTORY_PROPERTIES, **extra_args)
             elif table.path == PARENT_CASE_TABLE:
-                cls.__insert_system_properties(table, PARENT_CASE_TABLE_PROPERTIES)
+                cls.__insert_system_properties(table, PARENT_CASE_TABLE_PROPERTIES, **extra_args)
 
     @classmethod
-    def __insert_system_properties(cls, table, properties, top=True):
+    def __insert_system_properties(cls, table, properties, top=True, **data):
```
Comment (NoahCarnahan) [inline corehq/apps/export/models/new.py]: The `data` arg could use a more descriptive name. Maybe `column_initialization_data` or something like that?

File: corehq/apps/export/models/new.py (hunk 7 above)
Comment (NoahCarnahan) [inline corehq/apps/export/models/new.py]: Yeah, this function feels a bit unintuitive to me as well. What if we did something like:

    
    # add update_domain() method to StockExportColumn
    # add update_nested_repeat_count() to RowNumberColumn

    column = existing_column or copy(static_column)
    try:
        column.update_domain(domain)
    except AttributeError:
        pass
    try:
        column.update_nested_repeat_count(nested_repeat_count)
    except AttributeError:
        pass

This code is obviously less concise, but I think it's probably easier to figure out what it does at first glance. Using `isinstance()` would also be an alternative to the try/except that might look a little cleaner.
    
(I also think `extra_initialization_data()` is modifying the static columns here, so we should probably copy them before adding them to the table)

Comment (benrudolph) [inline corehq/apps/export/models/new.py]: nice, in a later commit i do copy the columns. i like that, i think i'll go with the `isinstance` so it's super clear that we're trying to update the row column and not any other column

File: corehq/apps/export/conversion_mappings.py [hunk 12]
```diff
@@ -15,7 +15,7 @@
     ("form.meta.@xmlns", None): ([PathNode(name='xmlns')], None),
     ("form.meta.appVersion.#text", None): ([PathNode(name='form'), PathNode(name='meta'), PathNode(name='appVersion')], None),
     ("form.meta.appVersion.@xmlns", None): ([PathNode(name='xmlns')], None),
-    ("form.case.@case_id", CASE_NAME_TRANSFORM): ([PathNode(name='form'), PathNode(name='case'), PathNode(name='@case_id')], CASE_NAME_TRANSFORM),
+    ("form.case.@case_id", "corehq.apps.export.transforms.case_id_to_case_name"): ([PathNode(name='form'), PathNode(name='case'), PathNode(name='@case_id')], CASE_NAME_TRANSFORM),
```
Comment (dimagimon) [inline corehq/apps/export/conversion_mappings.py]: E501 line too long (179 > 115 characters)

File: corehq/apps/export/static/export/js/models.js [hunk 13]
```diff
@@ -17,9 +17,9 @@ Exports.ViewModels.ExportInstance.prototype.getFormatOptionText = function(forma
     } else if (format === Exports.Constants.EXPORT_FORMATS.CSV) {
         return gettext('CSV (Zip file)');
     } else if (format === Exports.Constants.EXPORT_FORMATS.XLS) {
-        return gettext('Excel 2007');
+        return gettext('Excel (older versions)');
     } else if (format === Exports.Constants.EXPORT_FORMATS.XLSX) {
-        return gettext('Web Page (Excel Dashboards)');
+        return gettext('Excel 2007');
```
Comment (dimagimon) [inline corehq/apps/export/static/export/js/models.js]: 'gettext' is not defined. (no-undef)

File: corehq/apps/export/static/export/js/models.js [hunk 14]
```diff
@@ -17,9 +17,9 @@ Exports.ViewModels.ExportInstance.prototype.getFormatOptionText = function(forma
     } else if (format === Exports.Constants.EXPORT_FORMATS.CSV) {
         return gettext('CSV (Zip file)');
     } else if (format === Exports.Constants.EXPORT_FORMATS.XLS) {
-        return gettext('Excel 2007');
+        return gettext('Excel (older versions)');
```
Comment (dimagimon) [inline corehq/apps/export/static/export/js/models.js]: 'gettext' is not defined. (no-undef)

File: corehq/apps/export/static/export/js/models.js (hunk 8 above)
Comment (dimagimon) [inline corehq/apps/export/static/export/js/models.js]: 'Exports' is not defined. (no-undef)

File: corehq/apps/export/models/new.py (hunk 11 above)
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: https://github.com/dimagi/commcare-hq/pull/11065/commits/556d0fceb94990411465417e7801122ebf6d60ef

Comment (NoahCarnahan) [inline corehq/apps/export/models/new.py]: nice :+1:

**Comment by NoahCarnahan** [comment]
Looks good. Do you want to fix the lint errors now or should I go ahead and merge?

**Comment by benrudolph** [comment]
the lint errors are just out of date/not relevant. i'm going to open this against master so tests run first

**Pull request closed without merging**